src/generic/numberparsing.h

```diff
@@ -119,160 +119,578 @@ static inline bool is_made_of_eight_digits_fast(const char *chars) {
+    0x9f644ae5, 0x873d5d9f, 0xa90cb506, 0x9a7f124,  0xc11ed6d,  0x8f1668c8,
+    0xf96e017d, 0x37c981dc, 0x85bbe253, 0x93956d74, 0x387ac8d1, 0x6997b05,
+    0x441fece3, 0xd527e81c, 0x8a71e223, 0xf6872d56, 0xb428f8ac, 0xe13336d7,
+    0xecc00246, 0x27f002d7, 0x31ec038d, 0x7e670471, 0xf0062c6,  0x52c07b78,
+    0xa7709a56, 0x88a66076, 0x6acff893, 0x583f6b8,  0xc3727a33, 0x744f18c0,
+    0x1162def0, 0x8addcb56, 0x6d953e2b, 0xc8fa8db6, 0x1d9c9892, 0x2503beb6,
+    0x2e44ae64, 0x5ceaecfe, 0x7425a83e, 0xd12f124e, 0x82bd6b70, 0x636cc64d,
+    0x3c47f7e0, 0x65acfaec, 0x7f1839a7, 0x1ede4811, 0x934aed0a, 0xf81da84d,
+    0x36251260, 0xc1d72b7c, 0xb24cf65b, 0xdee033f2, 0x169840ef, 0x8e1f2895,
+    0xf1a6f2ba, 0xae10af69, 0xacca6da1, 0x17fd090a, 0xddfc4b4c, 0x4abdaf10,
+    0x9d6d1ad4, 0x84c86189, 0x32fd3cf5, 0x3fbc8c33, 0xfabaf3f,  0x29cb4d87,
+    0x743e20e9, 0x914da924, 0x1ad089b6, 0xa184ac24, 0xc9e5d72d, 0x7e2fa67c,
+    0xddbb901b, 0x552a7422, 0xd53a8895, 0x8a892aba, 0x2d2b7569, 0x9c3b2962,
+    0x8349f3ba, 0x241c70a9, 0xed238cd3, 0xf4363804, 0xb143c605, 0xdd94b786,
+    0xca7cf2b4, 0xfd1c2f61, 0xbc633b39, 0xd5be0503, 0x4b2d8644, 0xddf8e7d6,
+    0xcabb90e5, 0x3d6a751f, 0xcc51267,  0x27fb2b80, 0xb1f9f660, 0x5e7873f8,
+    0xdb0b487b, 0x91ce1a9a, 0x7641a140, 0xa9e904c8, 0x546345fa, 0xa97c1779,
+    0x49ed8eab, 0x5c68f256, 0x73832eec, 0xc831fd53, 0xba3e7ca8, 0x28ce1bd2,
+    0x7980d163, 0xd7e105bc, 0x8dd9472b, 0xb14f98f6, 0x6ed1bf9a, 0xa862f80,
+    0xcd27bb61, 0x8038d51c, 0xe0470a63, 0x1858ccfc, 0xf37801e,  0xd3056025,
+    0x47c6b82e, 0x4cdc331d, 0xe0133fe4, 0x58180fdd, 0x570f09ea};
 
-// called by parse_number when we know that the output is a float,
-// but where there might be some integer overflow. The trick here is to
-// parse using floats from the start.
-// Do not call this function directly as it skips some of the checks from
-// parse_number
-//
-// This function will almost never be called!!!
-//
-// Note: a redesign could avoid this function entirely.
-//
-static never_inline bool parse_float(const uint8_t *const buf, ParsedJson &pj,
-                                     const uint32_t offset, bool found_minus) {
-  const char *p = reinterpret_cast<const char *>(buf + offset);
-  bool negative = false;
-  if (found_minus) {
-    ++p;
-    negative = true;
+// handle case where strtod finds an invalid number. won't we have a buffer
+// overflow if it's just numbers past the end?
+// the assumption is that power_index is in the [0, 2*308] interval
+static really_inline double compute_float_64(uint64_t power_index, uint64_t i,
+                                             bool negative, bool *success) {
+  components c =
+      power_of_ten_components[power_index]; // safe because power_index <= 2*308
+  uint64_t factor_mantissa = c.mantissa;
+  int lz = leading_zeroes(i);
+  i <<= lz;
+  __uint128_t large_mantissa = (__uint128_t)i * factor_mantissa;
+  uint64_t upper = large_mantissa >> 64;
+  if (unlikely((upper & 0x1FF) == 0x1FF)) {
```

Comment:
we have a result in the higher order 64 bits, `upper`, and we want to determine if adding `i` to it (as per the math mentioned in the comments for this PR) will change the most significant 54 bits. since `i` is a `uint64_t` and we're looking at the most significant 64 bits, `i` < `1 << 64`, so if adding 1 to upper, which is equivalent to `1 << 64` in the original arithmetic, doesn't change the 54 bits, then we're ok. the only way adding 1 could change those bits is if we had a long enough string of trailing 1's at the end of `upper`

Reply:
also, we know that the leading 1 of `upper` will be either in most significant place, or the second most significant, because we've shifted `i` to have its leading 1 in its MSB and `f` as per the caches always has its leading 1 in its MSB. so, `2 ** 63 <= i, f <= 2 ** 64 - 1`. `(2 ** 63) ** 2 <= i * f <= (2 ** 64 - 1) ** 2`, thus `2 ** 126 <= i * f <= 2 ** 128 - ... < 2 ** 128`. i.e., it has the leading bit either 126 or 127 bits ahead of the start of the number

Reply:
could we have an additional check to see if the MSB is 1 and then use `0x3FF` instead of `0x1FF`? yes, but i found that in all those cases, it didn't affect the result. perhaps there was a bug in my implementation, but also it may be a mathematical connection i'm not seeing